crypto/rand/rand_crng_test.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * Implementation of the FIPS 140-2 section 4.9.2 Conditional Tests.
+ */
+
+#include <string.h>
+#include "internal/rand_int.h"
+#include "internal/thread_once.h"
+#include "rand_lcl.h"
+
+static RAND_POOL *crngt_pool;
+static unsigned char *crngt_prev;
+
+int (*crngt_get_entropy)(unsigned char *) = &rand_crngt_get_entropy_cb;
+
+int rand_crngt_get_entropy_cb(unsigned char *buf)
+{
+    size_t n;
+    unsigned char *p;
+
+    while ((n = rand_pool_acquire_entropy(crngt_pool)) != 0)
+        if (n >= CRNGT_BUFSIZ) {
+            p = rand_pool_detach(crngt_pool);
+            memcpy(crngt_prev, p, CRNGT_BUFSIZ);
+            rand_pool_reattach(crngt_pool, p);
+            return 1;
+        }
+    return 0;
+
+}
+void rand_crngt_cleanup(void)
+{
+    rand_pool_free(crngt_pool);
+    OPENSSL_secure_free(crngt_prev);
+    crngt_pool = NULL;
+    crngt_prev = NULL;
+}
+
+int rand_crngt_init(void)
+{
+    if ((crngt_pool = rand_pool_new(0, CRNGT_BUFSIZ, CRNGT_BUFSIZ)) == NULL)
+        return 0;
+    if ((crngt_prev = OPENSSL_secure_malloc(CRNGT_BUFSIZ)) != NULL
```

Comment:
There are very few places where the runtime calls secure malloc on its own.  Do ```git grep secure_malloc crypto``` to see.  Are we sure we want to do this? And are we sure that we need to do this?

Reply:
Entropy pools use secure memory and this is at least equivalent in importance (it contains the same seed material and it persists far longer).  They allocate using secure zalloc and don't appear in the grep.

There is, IMO, a better way.  Instead of storing the raw seed material for the comparison, I'd like to store a digest of it.  I asked the lab if this is allowed: their reading of the standard indicated a clear _no_.  They have requested clarification from NIST.

If this turns out to be acceptable, I'll switch to digesting the seed material and lose the secure memory.  A ruling might be a while coming though.